### PR TITLE
Add virtio-pmem support

### LINKED ISSUE #68: Add support for virtio-pmem
We need to add support for `virtio-pmem` if we want to expose some persistent memory to the guest, without using any ACPI.

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request introduces the support for virtio-pmem, which allows the VMM to expose some persistent memory to the guest, that can be backed by any type of file.

Fixes #68

### CODE REVIEW

**Comment by sboeuf** [comment]
Just a note here, in order to test I'm using a specific kernel branch https://github.com/sboeuf/linux/tree/virtio-pmem including the kernel patches adding the `virtio-pmem` support.

**Comment by sboeuf** [comment]
@sameo @chao-p PR updated with integration tests. PTAL
Don't be afraid by the `+3309` lines, they're mostly related to the new kernel config file I had to create for the kernel supporting the virtio-pmem patches :)

**Comment by sboeuf** [comment]
PR updated (rebased on the freshly merged `virtio-fs` code).

**Comment by richardwyang** [comment]
virtio-pmem kernel driver is not in current linux kernel, right?

BTW, how you verify this function in "guest"? Do we have ndctl for Clear Linux?

File: vm-virtio/src/pmem.rs [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+    size: u64,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for VirtioPmemConfig {}
+
+#[derive(Copy, Clone, Debug, Default)]
+#[repr(C)]
+struct VirtioPmemReq {
+    type_: u32,
```
Comment (rbradford) [inline vm-virtio/src/pmem.rs]: FYI, `r#type` is also cool syntax to solve this problem.

File: vmm/src/vm.rs [hunk 2]
```diff
@@ -620,6 +632,70 @@ impl DeviceManager {
...
+                    .open(pmem_cfg.file)
+                    .map_err(DeviceManagerError::PmemFileOpen)?;
+
+                if set_len {
+                    file.set_len(size)
+                        .map_err(DeviceManagerError::PmemFileSetLen)?;
+                }
+
+                let addr = unsafe {
+                    libc::mmap(
```
Comment (rbradford) [inline vmm/src/vm.rs]: Do we need to handle mmap failing?

**Comment by rbradford** [comment]
Long term it might be nice to add support for virtio-pmem to the firmware and then we'd also need to ensure we support just using the virtio-pmem as the root.

**Pull request merged**